legacy/logclient/persistance.go

```diff
@@ -0,0 +1,130 @@
+type QueryMeasurement struct {
+	startTime  time.Time
+	recorded   bool
+	numQueries uint64
+}
+
+// Debug only logs the given message if the verboseLogging option is true.
+func Debug(args ...interface{}) {
+	if PersistentAuditInfo.verboseLogging {
+		logrus.Debug(args...)
+	}
+}
+
+// CountGCRQuery adds the given number to the persistent audit logs only
+// if verboseLogging is enabled.
+func CountGCRQuery(n int) {
+	// Only capture queries when verboseLogging is enabled
+	if !PersistentAuditInfo.verboseLogging {
+		return
+	}
+	// Grab and the lock to prevent race conditions.
+	PersistentAuditInfo.mutex.Lock()
+	defer PersistentAuditInfo.mutex.Unlock()
+	// Determin if the current measurement has already been recorded.
+	if !PersistentAuditInfo.measurement.recorded {
+		PersistentAuditInfo.measurement.numQueries += uint64(n)
+		return
+	}
+	// Create a new measurement.
+	PersistentAuditInfo.measurement = QueryMeasurement{
+		startTime:  time.Now(),
+		recorded:   false,
+		numQueries: uint64(n),
+	}
+}
+
+// queryLogger records the number of queries within the specified
+// measurement window.
+func queryLogger() {
+	// This logger must not terminate unless the program exists.
+	for {
+		// Wait for the query window.
+		time.Sleep(GCRMeasurementWindow)
+		// Grab the number of queries, and mark as recorded.
+		var numQueries uint64
+		PersistentAuditInfo.mutex.Lock()
+		numQueries = PersistentAuditInfo.measurement.numQueries
+		PersistentAuditInfo.measurement.recorded = true
+		PersistentAuditInfo.mutex.Unlock()
+		// Log the number of queries.
+		msg := fmt.Sprintf("Recorded %d Queries to GCR within %d min.", numQueries, GCRMeasurementWindow/time.Minute)
+		Debug(msg)
+		// Record when the auditor surpasses the GCR query limit.
+		if numQueries > GCRQueryLimit {
+			msg = fmt.Sprintf("Quota limit %d req every %d min surpassed. Number of queries recorded: %d", GCRQueryLimit, GCRMeasurementWindow/time.Minute, numQueries)
+			Debug(msg)
+		}
+	}
+}
+
+// EnablePersistentLogging configures the persistent logger to capture and
+// and display logging extra information.
+func EnablePersistentLogging() {
+	// Run the logger in the background.
+	go queryLogger()
+	// Allow debugging statements.
+	PersistentAuditInfo.verboseLogging = true
+}
+
+const (
+	// GCRMeasurementQindow specifies the length of time to batch queries to GCR. Since Google's
+	// container registry specifies a query limit of 50,000 HTTP requests per 10 min, the window
+	// for recording queries is set to 10 min.
+	// Source: https://cloud.google.com/container-registry/quotas
+	GCRMeasurementWindow = time.Minute * 10
+	// GCRQueryLimit is the number of HTTP requests available to a single IP address within the
+	// specified GCRQueryWindow. For GCR, this is 50,000 HTTP requests.
+	// Source: https://cloud.google.com/container-registry/quotas
+	GCRQueryLimit = 50000
+)
+
+// PersistentAuditInfo is used to hold persistent logs for the auditor.
```

Comment:
I don't like this name `PersistentAuditInfo`. The word `Persistent` doesn't mean much here and is confusing (usually persistence means persisting the data to a lifecycle that exceeds the current duration of the program, such as adding an entry into a database or some other external system).